### PR TITLE
added optional labelText and semanticLabel to Checkbox

### PR DESCRIPTION
**Pull request by harperl-lgtm**
This PR added optional labelText (which will be  rendered side by side with Checkbox in the future, and thus is also announced by default by screen readers), and semanticLabel(which will be announced by screen reader if provided, overrides labelText, in order to do that we might want to wrap the Text widget inside ExcludeSemantics in the future).

Issues fixed:
[b/239564167]

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/checkbox.dart [hunk 1]
```diff
@@ -341,6 +343,18 @@ class Checkbox extends StatefulWidget {
...
 
+  /// The label text that will be rendered side by side with the [Checkbox].
+  /// Serves as default semantics label for the [Checkbox] if provided.
+  ///
+  // TODO(harperl-lgtm): Actually render the labelText in build method.
+  final String? labelText;
+
+  /// The semantics label that will be announced by a screen reader.
+  ///
+  /// Overrides labelText in terms of semantic labeling of the [Checkbox]
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/checkbox.dart:354]:
```suggestion
  /// Semantic label for the checkbox.
  ///
  /// Announced in accessibility modes (e.g TalkBack/VoiceOver).
  /// This label does not show in the UI.
  ///
  /// Defaults to [labelText], if specified.
```

File: packages/flutter/lib/src/material/checkbox.dart [hunk 2]
```diff
@@ -88,6 +88,8 @@ class Checkbox extends StatefulWidget {
     this.shape,
     this.side,
     this.isError = false,
+    this.labelText,
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/checkbox.dart:91]: Let's avoid adding an unused property and save that for a future PR

**Review by guidezpl — CHANGES_REQUESTED** [review]

**Review by guidezpl — APPROVED** [review]

**Comment by guidezpl** [comment]
@darrenaustin Was curious about your thoughts on the proposed TODO for this. `CheckboxListTile` is an overloaded widget for the simple case where text has to appear before or after a checkbox (and clicking that text toggles the checkbox).

File: packages/flutter/lib/src/material/checkbox.dart [hunk 3]
```diff
@@ -341,6 +342,20 @@ class Checkbox extends StatefulWidget {
...
+  //
+  // TODO(harperl-lgtm): Add to constructor parameters and actually render it.
+  // final String? labelText;
+
+  /// The semantic label for the checkobox that will be announced by screen readers.
+  ///
+  /// Announced in accessibility modes (e.g TalkBack/VoiceOver).
+  /// This label does not show in the UI.
+  ///
+  /// Defaults to labelText, if specified.
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/checkbox.dart:356]: And again, until the `labelText` is added, we should remove this reference.

**Review by darrenaustin — COMMENTED** [review]
LGTM, with a couple of comments below. If an app already has a label that they have added themselves and wrapped the check box and label in a `MergeSemantics`, what happens with this new `semanticLabel`?

**Comment by darrenaustin** [comment]
> @darrenaustin Was curious about your thoughts on the proposed TODO for this. `CheckboxListTile` is an overloaded widget for the simple case where text has to appear before or after a checkbox (and clicking that text toggles the checkbox).

It sounds reasonable to me, but perhaps there was a reason we chose not to implement it with a label. @HansMuller do you have any opinion/context on this?

**Comment by HansMuller** [comment]
> > @darrenaustin Was curious about your thoughts on the proposed TODO for this. `CheckboxListTile` is an overloaded widget for the simple case where text has to appear before or after a checkbox (and clicking that text toggles the checkbox).
> 
> It sounds reasonable to me, but perhaps there was a reason we chose not to implement it with a label. @HansMuller do you have any opinion/context on this?

Long ago we decided not to make intrinsic labels a feature of components like Checkbox. Over time that gap has been filled in - for better or worse - by the FooListTile components.  If the ListTile component is inconvenient for a common case, then we should make it convenient, perhaps with a named constructor.

File: packages/flutter/lib/src/material/checkbox.dart [hunk 4]
```diff
@@ -341,6 +342,20 @@ class Checkbox extends StatefulWidget {
   /// Must not be null. Defaults to false.
   final bool isError;
 
+  // The label text that will be rendered side by side with the [Checkbox].
+  // Serves as default semantics label for the [Checkbox] if provided.
+  //
+  // TODO(harperl-lgtm): Add to constructor parameters and actually render it.
+  // final String? labelText;
```
[the remark this answers is not in the document]
Comment (harperl-lgtm) [reply packages/flutter/lib/src/material/checkbox.dart]: Thank you so much for looking at my PR Darren!! :D labelText removed.

File: packages/flutter/lib/src/material/checkbox.dart (hunk 3 above)
Comment (harperl-lgtm) [reply packages/flutter/lib/src/material/checkbox.dart]: Thanks! Reference removed.

**Comment by harperl-lgtm** [comment]
> 

Thank you so much Hans for your insights! :D I agree it makes sense to pass a semantic label to a named constructor. Added in new commit!

**Comment by HansMuller** [comment]
This looks good. Are you going to add a `semanticLabel` property to CheckboxListTile as well?

**Comment by HansMuller** [comment]
I thought you'd just add a single semanticLabel constructor parameter to CheckboxListTile, I don't think the named constructor is needed (unless I'm missing something).  I realize that I brought up FooListTile named constructors in https://github.com/flutter/flutter/pull/116551#issuecomment-1353764955, but that suggestion assumed that an (unspecifed) common case was inconvenient.  In this case, it seems like just passing the semanticLabel parameter through, from the CheckboxListTile to its Checkbox, would be sufficient and convenient.

**Comment by harperl-lgtm** [comment]
Looping in @hangyujin who is going to take over some of my work. And thank you Hangyu for taking care of my PRs :)!

**Pull request closed without merging**